p2p: skip querying dns seeds if `-onlynet` disables IPv4 and IPv6

mzumsande opened the pull request:
Currently, `-onlynet` does not work well in connection with initial peer discovery, because DNS seeds only resolve to IPv6 and IPv4 adresses:
With `-onlynet=i2p`, we would load clearnet addresses from DNS seeds into addrman, be content our addrman isn't empty so we don't try to query hardcoded seeds (although these exist for i2p!), and never attempt to make an automatic outbound connection.
With `-onlynet=onion` and `-proxy` set, we wouldn't load addresses via DNS, but will make AddrFetch connections (through a tor exit node) to a random clearnet peer the DNS seed resolves to (see https://github.com/bitcoin/bitcoin/issues/6808#issuecomment-147652505), thus breaching the `-onlynet` preference of the user - this has been reported in the two issues listed below.

This PR proposes two changes:
1.) Don't load addresses that are unreachable (so that we wouldn't connect to them) into addrman. This is already the case for addresses received via p2p addr messages, this PR implements the same for addresses received from DNS seeds and fixed seeds. This means that in the case of `-onlynet=onion`, we wouldn't load fixed seed IPv4 addresses into addrman, only the onion ones.
2.) Skip trying the DNS seeds if neither IPv4 nor IPv6 are reachable and move directly to adding the hardcoded seeds from networks we can connect to. This is done by soft-setting `-dnsseed` to 0 in this case, unless `-dnsseed=1` was explicitly specified, in which case we abort with an `InitError`.

Fixes #6808
Fixes #12344

mzumsande committed:
p2p: add only reachable addresses to addrman

We will not make outgoing connection to peers that are unreachable
(e.g. because of -onlynet configuration).
Therefore, it makes no sense to add them to addrman in the first place.
While this is already the case for addresses received via p2p addr
messages, this commit does the same for addresses received
from fixed seeds.

mzumsande committed:
p2p: Don't query DNS seeds when both IPv4 and IPv6 are unreachable

This happens, for example, if the user specified -onlynet=onion or
-onlynet=i2p. DNS seeds only resolve to IPv4 / IPv6 addresses,
making their answers useless to us, since we don't want to make
connections to these.
If, within the DNS seed thread, we'd instead do fallback AddrFetch
connections to one of the clearnet addresses the DNS seed resolves to,
we might get usable addresses from other networks
if lucky, but would be violating our -onlynet user preference
in doing so.

Therefore, in this case it is better to rely on fixed seeds for networks we
want to connect to.

Co-authored-by: Vasil Dimov <vd@FreeBSD.org>

brunoerg reviewed (commented):
Strong concept ACK

luke-jr commented on test/functional/feature_config_args.py:
```diff
@@ -186,12 +186,28 @@ def test_seed_peers(self):
         assert time.time() - start < 60
         self.stop_node(0)
 
+        # No peers.dat exists, -dnsseed=1 but -onlynet=i2p
```
If `dnsseed=1` is set, I would expect it to use DNS seeds no matter what the other options are...

mzumsande replied:
It's the default value, so many users will have it set without explicitely choosing it. Or do you suggest to have behavior conditional on whether it is set as a default value, or explicitly specified by the user?

But I can't see a point in `-dnsseed=1 -onlynet=i2p` - for me, it is essentially like `-noconnect`: no attempts for outgoing connections are made, because we'll load address from DNS seeds that we then refuse to connect to.

luke-jr replied:
>It's the default value, so many users will have it set without explicitely choosing it. Or do you suggest to have behavior conditional on whether it is set as a default value, or explicitly specified by the user?

Yes, like SoftSetArg already does for other similar situations

>But I can't see a point in -dnsseed=1 -onlynet=i2p - for me, it is essentially like -noconnect: no attempts for outgoing connections are made, because we'll load address from DNS seeds that we then refuse to connect to.

Perhaps not, but for tor it does make some sense.

mzumsande replied:
> Yes, like SoftSetArg already does for other similar situations

Ok, I understand it would have to be separate from the other SoftSetArg interactions, because the precondition onlynet is only evaluated during `AppInitMain`

> Perhaps not, but for tor it does make some sense.

Why? I tend to see the current behavior for Tor (adding the dns seed as an [addr fetch](https://github.com/bitcoin/bitcoin/blob/a6fc293c0a1f27ba1e573bfa16fd76d5f58988b2/src/net.cpp#L1457-L1459), and making a clearnet-connection to a random node it resolves to) as a privacy bug and not something a user who specified `-onlynet=onion` could possibly want to do on purpose.

sipa replied:
It's undesirable to just ignore the user's request if they explicitly specify `-dnsseed=1`. Perhaps it makes sense to give an InitError, or even just a warning?

mzumsande replied:
Since we'd ignore a user's request whatever we'd to in this situation (either `-dnseed=1` or `-onlynet=onion`) I like the suggestion of returning an InitError and not ignoring either of the two conflicting requests. Will adjust the PR to implement that.

mzumsande replied:
Done with the latest push.

michaelfolkson commented:
Concept ACK. This was asked about in this StackExchange [question](https://bitcoin.stackexchange.com/questions/114781/dns-seeds-for-torv3).

sipa commented:
Concept ACK on (at least by default) not using DNS seed lookups (incl. addrfetch based ones) with `-onlynet=onion`. This is really a contradiction. Onlynet=onion means only connections to hidden services, and DNS seeds fundamentally requires making a connection to the IPv4IPv6 world (either by asking DNS servers, or by asking a Tor exit node to make a connection to what a DNS result resolves to).

Orthogonal idea: would it make sense, when using the hardcoded seeds (for whatever, onlynot or not), to only make addrfetch connections to them? Their point is getting us an entrypoint to the network, and using them directly as fully-featured nodes probably places an undue burden on them. This would get worse with a change like this PR.

mzumsande commented:
[09384fc ](https://github.com/bitcoin/bitcoin/commit/09384fc611c5b425aef3a1365cb322e12bbb73fa)to [1e039d1](https://github.com/bitcoin/bitcoin/commit/1e039d14b86a7c3c014f9302c159a21c302c6ccb): Addressed feedback, some refactoring/rewording.

The current PR is not specific to `-onlynet=onion` but affects all alternative networks. While I discussed Tor and I2P in the OP, is there anything special to CJDNS that would require another approach (ping @vasild)?

> Orthogonal idea (not for this PR): would it make sense, when using the hardcoded seeds (for whatever reason, onlynot or not), to only make addrfetch connections to them? Their point is getting us an entrypoint to the network, and using them directly as fully-featured nodes probably places an undue burden on them. This would get worse with a change like this PR.

Yes, I agree that this could makes sense. While in principle, the maximum inbound limit together with the inbound eviction logic should keep the number of connections under control,  hardcoded peers will receive connections from a lot of new nodes which require IBD, so it would be nice to reduce their traffic burden.

vasild commented:
Concept ACK

> is there anything special to CJDNS that would require another approach

Well, unlike Tor and I2P addresses, CJDNS addresses could be returned by the DNS seeds, e.g. this could happen:

```
seed.bitcoin.sipa.be has IPv6 address fc:fb1:4:387e:9248:9aff:febe:1c39
```

But we would still need to open a connection to the DNS server itself. If that is from the CJDNS network, then we are perfectly good (e.g. `/etc/resolv.conf` contains something like `nameserver fc:123...`). But if the DNS server is IPv4 or non-cjdns-IPv6, then opening a connection to it violates `onlynet=cjdns`, right? My understanding is that there is no easy and reliable way to check the address(es) of the DNS server(s) in use by the machine. They could also be changed after we have checked.

I guess, given the above, it is best and easiest to treat CJDNS like Tor and I2P and have `onlynet=cjdns` incompatible with `dnsseed=1`.

achow101 commented:
ACK 1e039d14b86a7c3c014f9302c159a21c302c6ccb

naumenkogs commented on src/net.cpp:
```diff
@@ -1635,9 +1638,13 @@ void CConnman::ThreadOpenConnections(const std::vector<std::string> connect)
 
             if (add_fixed_seeds_now) {
+                std::vector<CAddress> seed_addr{ConvertSeeds(Params().FixedSeeds())};
+                seed_addr.erase(std::remove_if(seed_addr.begin(), seed_addr.end(),
```
Wondering if this deserves a comment? For example, for a confused users which had this working previously.

naumenkogs replied:
Oops, I meant a Log statement, not a comment.

mzumsande replied:
I added a log "Added XXX fixed seeds from reachable networks", that would make it clear, I also think it might be of interest how many addresses were loaded (there is a similar log for the DNS seeds).

naumenkogs commented:
utACK 1e039d14b86a7c3c014f9302c159a21c302c6ccb

vasild reviewed (approved):
ACK 1e039d14b86a7c3c014f9302c159a21c302c6ccb

This would leave `gArgs.GetBoolArg("-dnsseed")` to `true` but would effectively disable DNS seeds if e.g. `-onlynet=onion` is given. The existent code is already capable of disabling DNS seeds - it does so if `dnsseed=0`.

Given that we want to disable DNS seeds, would it be simpler to flip (soft set) `-dnsseed` to `false` in `InitParameterInteraction()` and not touch any of the code in `net.{cpp.h}` (no need to add new member to `CConnman`)?

<details>
<summary>Would require something like this in InitParameterInteraction().</summary>

```cpp
bool clearnet_reachable = true;
if (args.IsArgSet("-onlynet")) {
    const auto onlynets = args.GetArgs("-onlynet");
    clearnet_reachable = std::any_of(onlynets.begin(), onlynets.end(), [](const auto& net) {
        const auto n = ParseNetwork(net);
        return n == NET_IPV4 || n == NET_IPV6;
    });
}
```
</details>

mzumsande commented on src/init.cpp:
```diff
@@ -1322,6 +1322,13 @@ bool AppInitMain(NodeContext& node, interfaces::BlockAndHeaderTipInfo* tip_info)
+    // Requesting DNS seeds requires connecting to IPv4/IPv6, which -onlynet options may prohibit:
+    // If -dnsseed=1 is explicitly specified, abort. If it's unspecified by the user, we don't request
+    // the DNS seeds per default but skip to the fixed seeds.
+    if (args.GetBoolArg("-dnsseed") == true && !IsReachable(NET_IPV4) && !IsReachable(NET_IPV6)) {
+        return InitError(strprintf(_("Incompatible options: -dnsseed was explicitly specified, but -onlynet forbids connections to IPv4/IPv6")));
+    };
```
There is already an interaction between `-forcednsseed` and `-dnsseed` which gives an InitError if `-dnsseed` is not set.
With the latest push (taking your suggestion to use `InitParameterInteraction()`) that should hit.

mzumsande commented on src/net.h:
```diff
@@ -1061,6 +1061,7 @@ class CConnman
     std::condition_variable condMsgProc;
     Mutex mutexMsgProc;
     std::atomic<bool> flagInterruptMsgProc{false};
+    std::atomic<bool> skip_dns_no_clearnet{false};
```
Obsolete because `skip_dns_no_clearnet` was removed with the latest push.

mzumsande commented:
> Given that we want to disable DNS seeds, would it be simpler to flip (soft set) `-dnsseed` to `false` in `InitParameterInteraction()` and not touch any of the code in `net.{cpp.h}` (no need to add new member to `CConnman`)?

Thanks - I agree that it would be simple and more similar to other parameter interactions to do that and took your suggestion with the latest push.

naumenkogs commented on src/net.cpp:
```diff
@@ -1635,10 +1638,15 @@ void CConnman::ThreadOpenConnections(const std::vector<std::string> connect)
             }
 
             if (add_fixed_seeds_now) {
+                std::vector<CAddress> seed_addr{ConvertSeeds(Params().FixedSeeds())};
```
nit: i think plural (`seed_addrs`) would make more sense, in the latter calls to this vector.

naumenkogs commented on src/init.cpp:
```diff
@@ -1322,6 +1332,13 @@ bool AppInitMain(NodeContext& node, interfaces::BlockAndHeaderTipInfo* tip_info)
+    // If -dnsseed=1 is explicitly specified, abort. If it's left unspecified by the user, we skip
+    // the DNS seeds by adjusting -dnsseed in InitParameterInteraction.
+    if (args.GetBoolArg("-dnsseed") == true && !IsReachable(NET_IPV4) && !IsReachable(NET_IPV6)) {
+        return InitError(strprintf(_("Incompatible options: -dnsseed was explicitly specified, but -onlynet forbids connections to IPv4/IPv6")));
```
nit: I think `specified` might as well mean `setting to false`, while you want to talk about `setting to true`

vasild replied:
Maybe just

```suggestion
        return InitError(strprintf(_("Incompatible options: -dnsseed=1 was explicitly specified, but -onlynet forbids connections to IPv4/IPv6")));
```

mzumsande replied:
Changed to `-dnsseed=1`

naumenkogs commented:
ACK 7dab71932c070896b68d2c780909c9ca817710f4

vasild reviewed (approved):
ACK 7dab71932c070896b68d2c780909c9ca817710f4

@naumenkogs's suggestions above make sense, would be happy to re-ACK.

jonatack commented on src/init.cpp:
```diff
@@ -719,6 +719,16 @@ void InitParameterInteraction(ArgsManager& args)
+            return n == NET_IPV4 || n == NET_IPV6;
+        });
+        if (!clearnet_reachable && args.SoftSetBoolArg("-dnsseed", false)) {
+            LogPrintf("%s: parameter interaction: -onlynet exludes IPv4 and IPv6 -> setting -dnsseed=0\n", __func__);
```
spelling: "excludes"

```suggestion
            LogPrintLevel(BCLog::ADDRMAN, BCLog::Level::Info, "%s: parameter interaction: -onlynet excludes IPv4 and IPv6 -> setting -dnsseed=0\n", __func__);
```

mzumsande replied:
fixed, thanks

jonatack commented on src/net.cpp:
```diff
@@ -1630,15 +1633,20 @@ void CConnman::ThreadOpenConnections(const std::vector<std::string> connect)
                 if (m_addr_fetches.empty() && m_added_nodes.empty()) {
                     add_fixed_seeds_now = true;
-                    LogPrintf("Adding fixed seeds as -dnsseed=0, -addnode is not provided and all -seednode(s) attempted\n");
+                    LogPrintf("Adding fixed seeds as -dnsseed=0 (or IPv4/IPv6 connections are disabled via -onlynet), -addnode is not provided and all -seednode(s) attempted\n");
```
while touching this, could use severity-based logging (info is always logged, like LogPrintf)
```suggestion
                    LogPrintLevel(BCLog::ADDRMAN, BCLog::Level::Info, "Adding fixed seeds as -dnsseed=0 (or IPv4/IPv6 connections are disabled via -onlynet), -addnode is not provided and all -seednode(s) attempted\n");
```

mzumsande replied:
I think I would prefer to have this done in a dedicated PR that switches it everywhere per module (for 25.0 - no idea whether this qualifies as a bugfix and should still go into 24.0, at least it still has the label), so that hopefully there wouldn't be too much of a mix for an extended time.

Also, how about a short heads-up in the weekly IRC meeting or so summarizing the recent changes? While logging is used by / affects everyone, I think most contributors who didn't follow the recent developments have internalized something like "With LogPrintf be careful about disk-filling attacks - with category-based logging no need to think about that much" which is going to change.

jonatack replied:
Yes agree, was about to update my comment to say never mind. (Am about to propose renaming the LogPrintLevel macro to simply Log or something before converting to more of them, as it is likely to subsume most of the other log macros.)

jonatack reviewed (commented):
Approach ACK, agree with @naumenkogs' suggestions as well

mzumsande commented:
[7dab719 ](https://github.com/bitcoin/bitcoin/commit/7dab71932c070896b68d2c780909c9ca817710f4)to [626ad77](https://github.com/bitcoin/bitcoin/commit/626ad77c36484b45bc158be575e5206b97268311): Addressed review feedback.

vasild reviewed (approved):
ACK 37b2b5bee84e1ae9678620fe2cdbdc3dd6b8608b

Sjors commented on src/net.cpp:
```diff
@@ -1475,6 +1475,9 @@ void CConnman::ThreadDNSAddressSeed()
                     found++;
                 }
+                vAdd.erase(std::remove_if(vAdd.begin(), vAdd.end(),
+                                          [](const CAddress& addr) { return !IsReachable(addr); }),
```
e91e3efac9c7b845fbabe30f23a1981706194ccb: why not just check this condition before calling `vAdd.push_back(addr)`? We should probably not increment `found` either when we don't use an address.

mzumsande replied:
Good point. On second thought, the filtering seems unnecessary in the DNS seed case, because with the second commit, we will skip querying the DNS seeds if IPv4 and IPv6 are unreachable, while DNS seeds only return IPv4 and IPv6 addresses. So I removed it.

Sjors commented:
Concept ACK

37b2b5bee84e1ae9678620fe2cdbdc3dd6b8608b ("don't query seeds...") makes sense to me, but e91e3efac9c7b845fbabe30f23a1981706194ccb ("add only reachable addresses to addrman") less so.

This commit seems to assume that the node is always started with the same `-onlynet` parameter, but what if the user decides to enable IPv4 / IPv6 later? `peers.dat` would be full of onion nodes in that case. Then again, eventually we'd query the DNS seeds again, so maybe it's not a big deal.

mzumsande commented:
> IIUC we only ever load the fixed seed nodes if `addrman.size() == 0`.

Yes, so we will never load them more than once for a given `peers.dat`.

> This commit seems to assume that the node is always started with the same `-onlynet` parameter, but what if the user decides to enable IPv4 / IPv6 later? `peers.dat` would be full of onion nodes in that case. Then again, eventually we'd query the DNS seeds again, so maybe it's not a big deal.
 
The general idea is to make the behavior consistent with the behavior towards peers (don't accept addrs into our addrman we can't or don't want to connect to).

I think if a user decides to just more additional networks (at least for a transitional period), there shouldn't be a problem: Both `GetAddr` answers and rumoured addresses from existing peer will likely contain a few entries from other networks over time, and once these are no longer unreachable due to `-onlynet` parameters, they will be accepted to addrman.

It's more complicated if the user makes an abrupt change from `-onlynet=X` to `-onlynet=Y`:
As you mentioned, switching to `-onlynet=IPv4/IPv6` should not be a problem, because we'll ask the DNS seeds even if `addrman.size() != 0` after 5 minutes.

I think that switching from `-onlynet=IPv4/IPv6` to e.g. `-onlynet=onion` or another network will probably not work as a general rule (irrespective of this PR): Since we likely used the DNS seeds to bootstrap, we have no IPs from previously unreachable networks in AddrMan, and we won't attempt the hardcoded seeds because addrman isn't empty.
 
What this PR may make more difficult is a situation where we switch between privacy networks, e.g. from `-onlynet=onion` to `-onlynet=i2p` because we may have queried the fixed seed and still have some random i2p fixed seeds in our addrman if we are lucky (and didn't overwrite these with others, which would likely happen after a while). This seems like a rather exotic situation to me, less likely than the previous one.

Maybe it would be better to change the criterion for querying hardcoded seeds from `addrman.size() == 0` to `# of reachable addresses == 0`, so that we would query the hardcoded seeds again if the existing entries all became unreachable due to a sudden change in `-onlynet` parameters?

Sjors commented:
`addrman.size() == 0 || reachable addresses == 0` sounds good to me

Sjors reviewed (commented):
utACK 91f0a7fbb79fe81a75370a4b60dcdd2e55edfa81 modulo using hardcoded seeds when we have nothing reachable

naumenkogs commented:
utACK 385f5a4c3feb716fcf3f2b4823535df6da6bb67b

vasild reviewed (approved):
ACK 385f5a4c3feb716fcf3f2b4823535df6da6bb67b

I experienced the problem discussed above (finding peers after `onlynet` changes) but forgot to fix or even report it back then. This discussion prompted me to report it at https://github.com/bitcoin/bitcoin/issues/26035 so that it does not get forgotten again.

Thanks!

kristapsk commented on src/net.cpp:
```diff
@@ -1635,10 +1635,15 @@ void CConnman::ThreadOpenConnections(const std::vector<std::string> connect)
             }
 
             if (add_fixed_seeds_now) {
+                std::vector<CAddress> seed_addrs{ConvertSeeds(Params().FixedSeeds())};
```
I agree with @Sjors on this.

fanquake replied:
Done in #26040.

Sjors commented:
Ok, so now the "modulo" part of my review is picked up in #26035.
